Sequence navs arrow keyboard support

alisan617 opened the pull request:
[TNL-5917](https://openedx.atlassian.net/browse/TNL-5917)

## Description
- Add arrow keyboard support to sequence navs

## Sandbox
[https://alisan617.sandbox.edx.org](https://alisan617.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/19a30717eff543078a5d94ae9d6c18a5/?child=first)

## Reviewers
- [ ] @cptvitamin 
- [ ] @andy-armstrong 
- [x] @bjacobel 
- [ ] @marcotuts 
- [ ] @catong  FYI

alisan617 commented:
jenkins run python

alisan617 commented:
jenkins run python

bjacobel commented:
Looks like this is branched off my branch, but there's some kind of merging issue where a commit of mine was duplicated.

alisan617 commented:
jenkins run lettuce

alisan617 commented:
jenkins run lettuce

alisan617 commented:
jenkins run lettuce

alisan617 commented:
@cptvitamin Per your request, I updated the sandbox with the `focus/active` class on arrow keypress.

alisan617 commented:
jenkins run js

alisan617 commented:
jenkins run bokchoy

alisan617 commented:
rebase off master, ready for review.

andy-armstrong commented on common/lib/xmodule/xmodule/js/spec/.gitignore:
```diff
@@ -7,3 +7,4 @@
 !time_spec.js
 !collapsible_spec.js
 !xmodule_spec.js
+!sequence/display_spec.js
```
I'm not sure why you want to Git ignore this file. I think that was only needed for CoffeeScript files so that the JS files didn't accidentally get checked in.

bjacobel replied:
The `!` means don't ignore, there's a rule higher up in this .gitignore that says to ignore `*.js` in this folder because it used to be all coffee.

andy-armstrong replied:
TIL. Thanks @bjacobel!

andy-armstrong commented on common/lib/xmodule/xmodule/js/src/sequence/display.js:
```diff
@@ -38,6 +38,12 @@
             this.displayTabTooltip = function(event) {
                 return Sequence.prototype.displayTabTooltip.apply(self, [event]);
             };
+            this.arrowKeys = {
```
Could this use HtmlUtils.constants? http://ui-toolkit.edx.org/utilities/constants/

bjacobel replied:
I would also like to see this, but I believe it isn't possible without requirejs because we don't expose `constants` on `window.edx`.

bjacobel commented on common/lib/xmodule/xmodule/js/spec/sequence/display_spec.js:
```diff
@@ -0,0 +1,104 @@
+                RIGHT: 39,
+                DOWN: 40
+            };
+        beforeEach(function() {
```
Could you add an empty line before `beforeEach`?

bjacobel commented on common/lib/xmodule/xmodule/js/spec/sequence/display_spec.js:
```diff
@@ -0,0 +1,104 @@
+        });
+
+        afterEach(function() {
+            delete window.XBlock;
```
There might be a better way to do this - after the last test has run, window.XBlock won't be defined any more, which could cause problems for other describe blocks or specs that follow this one.

bjacobel commented on common/lib/xmodule/xmodule/js/spec/sequence/display_spec.js:
```diff
@@ -0,0 +1,104 @@
+            var event = document.createEvent('Event');
+            event.keyCode = key;
+            event.initEvent('keydown', false, false);
+            document.dispatchEvent(event);
```
:thumbsup: for this - I was going to suggest using `new Event()` but it turns out it doesn't work on IE, so this is the way to go.

bjacobel commented on common/lib/xmodule/xmodule/js/src/sequence/display.js:
```diff
@@ -62,12 +69,65 @@
+                    self.nextNav($focused, index, total);
+                    break;
+
+                default:
```
Empty switch default.

alisan617 commented:
Note: I am addressing reviewers' comments. However, this PR will be on-hold, subject to a11y review.

_From @cptvitamin offline,_ the latest version of the spec recommends a few different keyboard behaviors then we have implemented here. I want to test them all out first before I make final recommendations.

cptvitamin commented:
@alisan617 this is looking great.  However, I want to be sure we are supporting all of the currently recommended best practices for this sort of widget (formerly known as a Tab Panel widget).  For reference:

Old guidance: (https://www.w3.org/TR/2013/WD-wai-aria-practices-20130307/#tabpanel)
New Guidance: (http://w3c.github.io/aria-practices/#tabpanel)

The core functionality of your implementation is excellent. Nice work!  However, I am suggesting the following change based on the latest guidance. 

The latest guidance contains a note that says:
>If the tab list is horizontal, it does not listen for Down Arrow or Up Arrow so those keys can provide their normal browser scrolling functions even when focus is inside the tab list.

I suggest removing the event listeners for the up and down arrow keys and only processing left and right.

That's it!

alisan617 commented:
@cptvitamin thank you for the suggestion!

@bjacobel comments addressed. @andy-armstrong please take a look when could.

alisan617 commented:
jenkins run bokchoy

alisan617 commented:
@bjacobel bump
